meta/include/gnuradio-4.0/meta/UncertainValue.hpp

```diff
@@ -0,0 +1,331 @@
+ * C/C++ Users Journal Volume 14, Number 3, March, 1996
+ * http://www.pennelynn.com/Documents/CUJ/HTML/14.03/MANNING/MANNING.HTM
+ *
+ * implements +,-,*,/ operators for basic arithmetic and complex types, for details see:
+ * https://en.wikipedia.org/wiki/Propagation_of_uncertainty#Example_formulae
+ * This implements only propagation of errors between variables that are uncorrelated.
+ * A more rigorous treatment would require the calculation and propagation of the
+ * corresponding covariance matrix which is out of scope of this implementation.
+ */
+
+template<typename T>
+concept arithmetic_or_complex_like = std::is_arithmetic_v<T> || complex_like<T>;
+
+template<arithmetic_or_complex_like T>
+struct UncertainValue {
+    using value_type = T;
+
+    T value       = static_cast<T>(0); /// mean value
+    T uncertainty = static_cast<T>(0); /// uncorrelated standard deviation
+
+    // Default constructor
+    constexpr UncertainValue() noexcept = default;
+
+    constexpr UncertainValue(T value_, T uncertainty_) noexcept : value(value_), uncertainty(uncertainty_) {}
+
+    explicit(false) constexpr UncertainValue(T value_) noexcept : value(value_), uncertainty(static_cast<T>(0)) {}
+
+    constexpr UncertainValue(const UncertainValue &) noexcept = default;
+    constexpr UncertainValue(UncertainValue &&) noexcept      = default;
+    constexpr UncertainValue &
+    operator=(const UncertainValue &) noexcept
+            = default;
+    constexpr UncertainValue &
+    operator=(UncertainValue &&) noexcept
+            = default;
+    ~UncertainValue() = default;
+
+    constexpr UncertainValue &
+    operator=(const T &other) noexcept {
+        value       = other;
+        uncertainty = static_cast<T>(0);
+        return *this;
+    }
+
+    constexpr UncertainValue &
+    operator=(T &&other) noexcept {
+        value       = std::move(other);
+        uncertainty = static_cast<T>(0);
+        return *this;
+    }
+
+    auto operator<=>(UncertainValue const&) const = default;
+};
+
+template<typename T>
+UncertainValue(T, T) -> UncertainValue<T>;
+
+template<typename T>
+concept UncertainValueLike = requires(T t) {
+    typename T::value_type;                                           // UncertainValue must have a nested value_type
+    { t.uncertainty } -> std::convertible_to<typename T::value_type>; // Must have an 'uncertainty' field
+};
+
+namespace detail {
+template<typename T>
+struct UncertainValueTypeType {
+    using type = T;
+};
+
+template<typename T>
+struct UncertainValueTypeType<UncertainValue<T>> {
+    using type = T;
+};
+} // namespace detail
+
+template<typename T>
+using UncertainValueType_t = detail::UncertainValueTypeType<T>::type;
+
+/********************** some basic math operation definitions *********************************/
+
+template<typename T, typename U>
+    requires UncertainValueLike<T> || UncertainValueLike<U>
```

Comment:
Instead of a concept-based check, I'd rather have a check that uses `is_instantiation_of`.

This way, if someone has a custom type with value and uncertainty fields in their type, this will create a operator+ for them quite unexpectedly. Potentially even collide with their type's operator+.